Breaking down a problem: Content (#2)
Ada-Developers-Academy/core-unit-1: approaching-a-problem/breaking-down-a-problem.md

# Breaking Down a Problem

## Learning Goals

- Describe steps used for breaking down a problem
- Define timeboxing

## Introduction

> Programming code is about using computers to solve problems. Often, this initial problem will be too complex for us to solve in one attempt. In order to begin solving these problems, we need to practice breaking down a problem.

Author: ```suggestion
Writing code is about using computers to solve problems. Often, this initial problem will be too complex for us to solve in one attempt. In order to begin solving these problems, we need to practice breaking down a problem.
```

Going with "writing"!

(I have this theory that "programming" is an awkward verb for folks not used to the word; programming in other fields can mean scheduling, the link that `writing code == programming` isn't quite there, etc etc etc)

## Vocabulary and Synonyms

| Vocab | Definition | Synonyms | How to Use in a Sentence |
| ----- | ---------- | -------- | ------------------------ |
Timeboxing | A time management technique where someone determines the maximum amount of time to spend on a task | - | "My code is working, but I want to make it better. I'll timebox refactoring to 2 hours," "I've been stuck on this and I need to do research. I'll timebox research to 30 minutes and then ask a classmate for help."

## Breaking Down a Problem

> The steps to breaking down a programming problem can be identical to real-life problem-solving. In order to come to express these steps, we will explain one angle of real-life problem-solving.

Author: ```suggestion
The steps to breaking down a programming problem can be identical to real-life problem-solving. To get used to thinking step-by-step, let's try breaking down a common situation.
```

Went with @anselrognlie 's suggestion, but switched the sentence order for motivation/goal first. Tagging ansel for visibility

### Real Example

Here's a problem that we might face often: **How do we prepare dinner for tonight?**

#### Ask Clarifying Questions

The first thing we can do is make sure we understand the problem. We can **clarify the problem with questions.** These questions and answers should give us constraints to the problem, making the problem (and its solution) smaller and clearer.

> Clarifying questions and their answers give us constraints. Clarifying questions:
> 
> 1. Make a problem (and its solution) smaller and clearer
> 2. Confirm or reject any assumptions

Author: ```suggestion
Clarifying questions helps us by:

1. Making a problem (and its solution) smaller and clearer
2. Confirming or rejecting any assumptions
```

Tagging @anselrognlie for visibility, hopefully this lowers the redundancy

## Always Multiple Solutions
We can realize here that **there are multiple ways to solve a problem.** We could have looked for a recipe that only has ingredients we already have. Or, we could have ordered pizza delivery!

### !end-callout

#### Come Up With Ideas

Once we have some sub-problems, we can start coming up with ideas for how to solve one of them. We should think through the solution with all of the resources available! This includes writing down ideas, notes, pseudocode, comments, and explanations. We should also make a list of tasks we need to do, topics to do more research on, or more questions we have.

For our example, we can start to think:

- Could we find a recipe in a recipe book? Or can we find one online? Or do we want to call a family member for a recipe?
- Can we make a list of ingredients we need to buy by comparing the recipe to our pantry?
- To get to the grocery store, can we take the train?
- We don't own a microwave. Can we use the oven instead?

#### Begin Solving One Sub-problem

Now that we have some ideas, we can pick one sub-problem and one strategy. Then, we begin making one possible solution. We might begin to solve a sub-problem like so:

> Go shopping for the ingredients we need

Reviewer: ```suggestion
Go shopping for the ingredients we need:
```
Added colon

  - Go to the train station
  - Take the train to the market
  - Browse every market stall for every ingredient
  - When we find an ingredient, buy it
  - Take the train home

#### Repeat As Necessary

We should **reflect** on our solution so far and determine if we're moving closer to a final solution. We can ask ourselves how to adjust if we need to.

Then, we should **repeat** these steps for different sub-problems until we have our final solution. We may find that our sub-problems have sub-problems! This process might continue until we are in a state that we comfortable with.

### Summarized Steps

1. Read the problem
2. Clarify the problem
3. Create sub-problems
4. Hypothesize ideas for the solution

> 5. Solve a sub-problem in Python code

Author: Sweeet!

I took out the two spaces between the sentences to conform with the surrounding writing style.

6. Repeat steps necessary

## Working First, Improve Second

Sometimes we're working on a solution. Then, suddenly, half-way through this solution, we think of a second solution! This second solution is so much better, cleaner, and more perfect! Do we begin to work on that solution now?

Instead of abandoning your first solution, programmers should try to finish an initial working solution first.

> Programmers should get their code working first, and then improve their code afterwards. When programmers focus on solving the problem first, and making their code elegant later, programmers can focus more and accomplish more!

Author: Hope you don't mind that I edited your suggestion;
- conform to one space between sentences
- modified `solving a problem!` to `solving a problem.`

## Timeboxing

Timeboxing is a time management technique where someone determines the _maximum_ amount of time to spend on a task.

When we are solving problems, we may come up with one solution, but then desire to work more on it. Or, sometimes, we might want to research alternative solutions. In these cases, it might be valuable to timebox these tasks.

Some examples might be:
- Timeboxing "Research alternatives to if/else conditionals" to 1 hour, and decide on one solution at that point
- Timeboxing "Refactor our project" to 2 hours, and call the project good enough at that point

> - Timeboxing "Work through the binary problems" and getting stuck for 30 minutes, and ask for help from the cohort at that point

Author: ```suggestion
- Timeboxing "Work through the binary problems solo" to 30 minutes, including using debugging strategies, then ask for help from the cohort if I'm stuck
```
